### PR TITLE
Add new DomainType for application usage

### PR DESCRIPTION
**Pull request by saolyn**
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**

This PR creates  the `DOMAIN_APPLICATION_MASK` DomainType for applications as defined in the Beacon Chain consensus spec.

**Other notes for review**

The value unfortunately had to be hardcoded to `[4]byte{0, 0, 0, 1}` as `Uint64ToBytesBigEndian(1)` returns 8 bytes and `ToBytes4` slices 4 leading 0's.

### CODE REVIEW

File: config/params/mainnet_config.go [hunk 1]
```diff
@@ -165,6 +165,7 @@ var mainnetBeaconConfig = &BeaconChainConfig{
 	DomainSyncCommittee:               bytesutil.ToBytes4(bytesutil.Bytes4(7)),
 	DomainSyncCommitteeSelectionProof: bytesutil.ToBytes4(bytesutil.Bytes4(8)),
 	DomainContributionAndProof:        bytesutil.ToBytes4(bytesutil.Bytes4(9)),
+	DomainApplicationMask:             [4]byte{0, 0, 0, 1},
```
Comment (terencechain) [inline config/params/mainnet_config.go:168]: Another option is `bytesutil.ToBytes4(common.HexToHash("0x00000001").Bytes()[28:])` but that's obvious uglier. Yours is good!

**Review by terencechain — APPROVED** [review]
Implements https://github.com/ethereum/consensus-specs/pull/2884
Lgtm!

File: config/params/mainnet_config.go (hunk 1 above)
Comment (nisdas) [reply config/params/mainnet_config.go:168]: This is a drop in readability imo, why can't we follow the previous pattern ?
```
bytesutil.ToBytes4(bytesutil.Bytes4(X)),
```

Comment (terencechain) [reply config/params/mainnet_config.go:168]: I see three options to this:

1.) bytesutil.ToBytes4(bytesutil.Bytes4(16777216))
2.) bytesutil.ToBytes4(common.HexToHash("0x00000001").Bytes()[28:])
3.) [4]byte{0, 0, 0, 1},

If we compare all three with the spec definition `DomainType('0x00000001'), 2 or 3 is the closest. One which we hard a magic number is not pretty, and it's less auditable. Any options that I may have missed?

Comment (nisdas) [reply config/params/mainnet_config.go:168]: My main concern is just reading how the config is initialized and trying to know why `[4]byte{0, 0, 0, 1}` was written the way it is. It isn't obvious until I read the spec again. Either you can add a comment stating that it is intended to be a mask which is why it was initialized that way. Or you can try doing `bytesutil.ToBytes4(bytesutil.ReverseByteOrder(bytesutil.Bytes4(1)))`. But then it is messier to read.

Comment (terencechain) [reply config/params/mainnet_config.go:168]: I'm ok with either option. Will let @saolyn decide

**Comment by nisdas** [comment]
Also tests are failing @saolyn

File: config/params/mainnet_config.go (hunk 1 above)
Comment (saolyn) [reply config/params/mainnet_config.go]: @nisdas, @terencechain, all of the proposed solutions were a bit messy.
As an alternative, I decided to create a new function `Uint32ToBytes4` that will do the same thing as `bytesutil.ToBytes4(bytesutil.Bytes4(N))` but will also work for other values. 
This way it should be a lot cleaner and better represent the actual values from the spec.

Please let me know what you think.

File: encoding/bytesutil/bytes.go [hunk 2]
```diff
@@ -339,6 +339,14 @@ func HighestBitIndexAt(b []byte, index int) (int, error) {
 	return 0, nil
 }
 
+// Uint32ToBytes4 is a convenience method for converting uint32 to a fix
+// sized 4 byte array in big endian order. Returns 4 byte array.
+func Uint32ToBytes4(i uint32) [4]byte {
```
[the remark this answers is not in the document]
Comment (saolyn) [reply encoding/bytesutil/bytes.go:344]: On it!

**Review by terencechain — APPROVED** [review]

**Pull request merged**